src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingSourceGenerator.Helpers.cs

```diff
@@ -0,0 +1,149 @@
+            public const string KeyTypeNotSupported = "The collection key type is not supported";
+            public const string MultiDimArraysNotSupported = "Multidimensional arrays are not supported.";
+            public const string NullableUnderlyingTypeNotSupported = "Nullable underlying type is not supported";
+            public const string TypeNotDetectedAsInput = "Generator parser did not detect the type as input";
+            public const string TypeNotSupported = "The type is not supported";
+        }
+
+        private static class TypeFullName
+        {
+            public const string Array = "System.Array";
+            public const string ConfigurationKeyNameAttribute = "Microsoft.Extensions.Configuration.ConfigurationKeyNameAttribute";
+            public const string Dictionary = "System.Collections.Generic.Dictionary`2";
+            public const string GenericIDictionary = "System.Collections.Generic.IDictionary`2";
+            public const string HashSet = "System.Collections.Generic.HashSet`1";
+            public const string ISet = "System.Collections.Generic.ISet`1";
+            public const string IConfigurationSection = "Microsoft.Extensions.Configuration.IConfigurationSection";
+            public const string IConfiguration = "Microsoft.Extensions.Configuration.IConfiguration";
+            public const string IDictionary = "System.Collections.Generic.IDictionary";
+            public const string IServiceCollection = "Microsoft.Extensions.DependencyInjection.IServiceCollection";
+            public const string List = "System.Collections.Generic.List`1";
+        }
+
+        private static bool TypesAreEqual(ITypeSymbol first, ITypeSymbol second)
+                => first.Equals(second, SymbolEqualityComparer.Default);
+
+        private enum InitializationKind
+        {
+            None = 0,
+            SimpleAssignment = 1,
+            AssignmentWithNullCheck = 2,
+            Declaration = 3,
+        }
+
+        private sealed class SourceWriter
+        {
+            private readonly StringBuilder _sb = new();
+            private int _indentationLevel;
+
+            public int Length => _sb.Length;
+            public int IndentationLevel => _indentationLevel;
+
+            public void WriteBlockStart(string declaration)
+            {
+                WriteLine(declaration);
+                WriteLine("{");
+                _indentationLevel++;
+            }
+
+            public void WriteBlockEnd(string? extra = null)
+            {
+                _indentationLevel--;
+                Debug.Assert(_indentationLevel > -1);
+                WriteLine($"}}{extra}");
+            }
+
+            public void WriteLine(string source)
+            {
+                string indentationSource = _indentationLevel == 0
+                    ? ""
+                    : string.Join("", Enumerable.Repeat(" ", 4 * _indentationLevel));
+
+                _sb.AppendLine($"{indentationSource}{source}");
```

Comment:
I think you could avoid quite a few intermediate allocations here:
```suggestion
                const string FourSpaces = "    ";
                for (int i = 0; i < _indentationLevel; i++)
                    _sb.Append(FourSpaces);

               _sb.AppendLine(source);
```

Reply:
Or you could use `_sb.Append(' ', 4 * _indentationLevel)`

Reply:
Thanks